packages/core/src/input/InputManager.ts

```diff
@@ -0,0 +1,358 @@
+              _pointerList[lastCount] = removedTouch;
+            }
+            _input.pageX = (_input.pageX * lastCount - removedTouch.pageX) / nowCount;
+            _input.pageY = (_input.pageY * lastCount - removedTouch.pageY) / nowCount;
+          }
+          delete _pointerIdToIndex[pointerId];
+        }
+        break;
+      case PointerChangeType.Update:
+        if (idx === undefined) {
+          if (lastCount > 0 && !this._multiTouchEnabled) {
+            return;
+          }
+          let touch: Pointer;
+          if (!_pointerList[lastCount]) {
+            touch = _pointerList[lastCount] = new Pointer(pointerId, pageX, pageY, lastCount);
+          } else {
+            touch = _pointerList[lastCount];
+            touch.pageX = pageX;
+            touch.pageY = pageY;
+            touch.indexInList = lastCount;
+          }
+          _pointerIdToIndex[pointerId] = lastCount;
+          const nowCount = (this._actPointerCount = lastCount + 1);
+          _input.pageX = (_input.pageX * lastCount + pageX) / nowCount;
+          _input.pageY = (_input.pageY * lastCount + pageY) / nowCount;
+        } else {
+          const touch = _pointerList[idx];
+          _input.pageX += (pageX - touch.pageX) / lastCount;
+          _input.pageY += (pageY - touch.pageY) / lastCount;
+          touch.pageX = pageX;
+          touch.pageY = pageY;
+        }
+        break;
+      default:
+        break;
+    }
+  }
+
+  /**
+   * Push event types into an ordered queue.
+   * @param type - Pointer event type
+   */
+  private _pushEventList(type: PointerEventType) {
+    const { _eventList } = this;
+    if (_eventList.length == this._eventLen) {
+      _eventList.push(type);
+    } else {
+      _eventList[this._eventLen] = type;
+    }
+    ++this._eventLen;
+  }
+
+  /**
+   * On pointer down.
+   * @param evt - Pointer Event
+   */
+  private _onPointerDown = (evt: PointerEvent) => {
+    this._changePointer(PointerChangeType.Update, evt.pointerId, evt.pageX, evt.pageY);
+    if (this._actPointerCount == 1) {
+      this._pushEventList(PointerEventType.PointDown);
+    }
+  };
+
+  /**
+   * On pointer up.
+   */
+  private _onPointerUp = () => {
+    if (this._actPointerCount == 1) {
+      this._pushEventList(PointerEventType.PointUp);
+    }
+  };
+
+  /**
+   * On pointer move.
+   * @param evt - Pointer Event
+   */
+  private _onPointerMove = (evt: PointerEvent) => {
+    this._changePointer(PointerChangeType.Update, evt.pointerId, evt.pageX, evt.pageY);
+  };
+
+  /**
+   * On pointer cancel or out.
+   * @param evt - Pointer Event
+   */
+  private _onPointerCancelOrOut = (evt: PointerEvent) => {
+    this._changePointer(PointerChangeType.Remove, evt.pointerId);
+  };
+
+  /**
+   * Update the current listening status.
+   */
+  private _updateListener() {
+    const { _regScriptCount, _hadListener, _canvas: canvas } = this;
+    if (_regScriptCount > 0 && !_hadListener) {
```

Comment:
`_hadListener` maybe no useful, `_regScriptCount ` is enough.

Reply:
Done.